### DISCUSSION TITLE
Restoring sessions for web3wallet

### DISCUSSION THREAD

**Discussion by jonandgon** [WalletConnect/walletconnect-monorepo#2574] [Developer Support] [answered]
The client-side signClient stores sessions when a dApp connects to a wallet. If the dApp closes and reopens, the sessions will still be accessible. 

As far as I can tell, this isn't the case for web3wallet. When the browser closes and the wallet dies, the sessions are not restored for the wallet. I have checked the localstorage for my wallet and it doesn't look like there is much being stored. 

Is there something I'm missing? Session restoration is vital for a good user experience.

**Comment by jonandgon** [comment]
update: https://github.com/orgs/WalletConnect/discussions/979 seems to be a related unresolved issue, but it was created before web3wallet came out.

**Reply by finessevanes** [reply]
Hey, you should be able to get it using `getPairings`: https://docs.walletconnect.com/2.0/specs/clients/core/pairing/pairing-api

**Reply by jonandgon** [reply]
The pairings/sessions aren't stored in local storage on the wallet side, so getPairings returns nothing after the wallet is closed and reopened.

**Reply by jonandgon** [reply]
For clarity, I expect similar functionality as SignClient for Web3Wallet, [as shown in your docs](https://docs.walletconnect.com/2.0/web/sign/dapp-usage#restoring-a-session).

The above docs mention that sessions are stored in local storage for SignClient, but sessions are not stored in local storage for Web3Wallet.

**Comment by afcharliekong** [comment]
Try `getActiveSessions()` ? Its under the `IWeb3Wallet`.

**Reply by jonandgon** [reply]
@afcharliekong no luck, It returns an empty array. I just double checked the local storage of my wallet after connecting my dapp to my wallet and nothing from walletconnect/web3wallet is there at all.

Here is a screenshot of my dApp's local storage after connecting. You can see that there's some walletconnect info being stored, and that info is grabbed each time I refresh the dApp.
<img width="314" alt="Screenshot 2023-06-19 at 1 31 27 PM" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/75697675/5f550159-6dd6-4b4f-be9f-042e0c28128e">


Here is what I see on my wallet's local storage after connecting with a dApp. There is nothing from walletconnect stored locally. The session info appears to only be stored in memory and is wiped each time the wallet closes.
<img width="632" alt="Screenshot 2023-06-19 at 1 32 00 PM" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/75697675/0d4ea48e-4ed8-498e-972e-7f26d4e89fa4">

**Reply by afcharliekong** [reply]
You should see something like this if you are connected to a Dapp. I'm filtering the wc related keys.

<img width="717" alt="image" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/103051707/e93192a1-10df-4aaf-9ddc-eafdcd0553c0">

Take a look at their example https://github.com/WalletConnect/web-examples/tree/main/wallets/react-wallet-v2 and see if there's any issue with your implementation?

**Comment by jonandgon** [comment]
Ahh I think I may have found the error. WalletConnect calls` window.localStorage` to save session data, but my web3wallet instance is running in the background script of a chrome extension. Background scripts don't have access to `window`, so they must use `chrome.storage.local` instead.

Is there any workaround that you can think of using? I could try to patch the walletconnect packages myself using pnpm, but this is not ideal in the long term.

**Reply by jonandgon** [reply]
I'd also be open to the idea of saving the walletconnect sessions in storage myself, however there is no function I'm aware of that allows a Web3Wallet instance to be initialized with existing session data.

**Reply by ScorpioEF** [reply, marked as answer]
In API Core, you can inject you own storage solution. So instead of using the default one you can use `chrome.storage.local`.

```
const core = new Core({
  projectId: 'something',
  storage: {
    async getKeys(): Promise<string[]> { /* code */ }
    async getEntries<T = any>(): Promise<[string, T][]> { /* code */ }
    async getItem<T = any>(key: string): Promise<T | undefined> { /* code */ }
    async setItem<T = any>(key: string, value: T): Promise<void> { /* code */ }
    async removeItem(key: string): Promise<void> { /* code */ }
  },
});

Web3Wallet.init({core, metadata});
```
